### DISCUSSION TITLE
(Next.js) Empty POST request body in API route, but only from certain user agents

### DISCUSSION THREAD

**Discussion by jasonwebb** [vercel/community#1521] [Help] [answered]
### Summary

I built an API route that is intended to be used by a Unity game to "sign in" players when they first load the game. The API route  accepts simple JSON data over POST and returns a status code and JSON message indicating whether the data matches a user in our database. 

It works perfectly for me when I interact with it in my main web app (an admin interface), and with third-party tools like Postman and ReqBin. But when the Unity app (built by another dev) sends a POST request, the API route returns one of my custom error messages saying that no variables were found. The weird part is that the Unity dev can send the exact same POST request to a Postman mock server and the body comes through perfectly fine!

### Additional information

To be clear, I don't expect any support around the Unity code. Let's assume that's set up correctly and try to eliminate the API route as the source of the problem. I'm really just trying to figure out where and why the request body is being dumped, which seems to only be happening with my API route in requests sent from the Unity app.

Here is what I've been able to learn so far:

1. From doing some console logging of the request object in the API route's handler function and watching the Function Logs, I can see that all the headers for the POST requests from the Unity app are coming through just fine ([see the full request object here](https://gist.github.com/jasonwebb/188767f24b3eab835c9d642a07047077)). It's just the body that is empty.
2. I've tried [disabling the body parser](https://nextjs.org/docs/api-routes/request-helpers#custom-config:~:text=export%20const%20config%20%3D%20%7B%0A%20%20api%3A%20%7B%0A%20%20%20%20bodyParser%3A%20false%2C%0A%20%20%7D%2C%0A%7D) and logging the request again, but the body is still empty.
3. The Unity app is able to interact with several other web APIs using the exact same code (using different data, of course), so we're confident it's set up right.
4. The Unity app is able to send the exact same payload to a Postman mock server, where it is received and parsed as expected.
5. The Unity app does receive responses from the API route with one of my custom error messages, which states that the API script did not find the variables it expected in the request body.
6. I am able to send JSON data to the API route from Postman and ReqBin and confirm that the route works as expected that way.
7. I've had the Unity dev send me the exact string that is attached to the HTTP POST request, and it appears valid (checked with a couple of JSON linters).

I have a suspicion that the JSON string being sent as the request's payload is slightly imperfect and can't be parsed by either Vercel or Next.js, but I can't seem to dig down far enough to see where that happens. Strange that Postman's mock server is able to parse the body though - maybe they some behind-the-scenes pre-processing / cleanup code?

**Comment by amyegan** [comment]
Hi Jason! Are you able to share a minimal repro or any example links so I can get a better look at what is going wrong?

**Reply by jasonwebb** [reply]
I'm actually not able to reproduce this myself at all, it seems to only happen with requests from Unity. I've even tried using Postman to recreate the issue by copying over the exact headers scraped from the requests and using the actual JSON string from Unity (provided to me by the Unity dev logging it to his local console). The requests work fine in Postman, but not when they come from Unity. The request headers come through OK in both cases, but the Unity requests always have an empty request body.

Here's an [example of the raw request](https://gist.github.com/jasonwebb/188767f24b3eab835c9d642a07047077) from Unity when my API route is essentially the following:

```
export const config = {
  api: {
    bodyParser: false,
  },
}

export default async function handler(req, res) {
  console.log(req);
  return res.status(200).json({ req.body });
}
```

It seems like either Vercel's servers or Next.js is dropping the request body by the time it reaches my app code, and I'm not sure how to dig deeper.

**Reply by jasonwebb** [reply]
Update: the Unity dev deployed a simplified test app to AWS that we can use to send and debug requests with. That app is here: https://rem5-unitywebgl-test.s3.us-east-2.amazonaws.com/API-Test/Simulacra_TEST_API_2023-02-14-002/index.html

Right away he noticed an error in the browser's console log pointing to a CORS issue, with didn't happen when the Unity app was being run locally (that we know of):
![image](https://user-images.githubusercontent.com/422783/218931155-84e341b4-ac89-45dd-9d82-9972458ad465.png)

So we started looking into enabling CORS on the API route, and found lots of different solutions. I just got done trying all the ones I could find, but in every instance I get this other CORS error when making requests through the Unity test app and running `fetch()` requests from various places like CodePen:

```
XMLHttpRequest at 'https://simulacra-admin-app.vercel.app/api/users/sign-in' from origin 'https://rem5-unitywebgl-test.s3.us-east-2.amazonaws.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

Here are all the things I've tried so far:

1. Adding a "headers" function to `next.config.js` [as shown in this Vercel doc](https://vercel.com/guides/how-to-enable-cors#enabling-cors-in-a-next.js-app). I tried it as-is and with the `Access-Control-Allow-Credentials` and `Access-Control-Allow-Headers` headers removed.
2. Changing the value for the `source` key in the above `next.config.js` function to `/api/(.*)`, and then to the exact API route I'm testing (`/api/users/sign-in`).
3. Using the `cors` package to set up CORS via middleware, as shown in the [official Vercel + Next.js example for API routes with CORS](https://github.com/vercel/next.js/tree/canary/examples/api-routes-cors).
4. Using the [`nextjs-cors` package](https://www.npmjs.com/package/nextjs-cors), which I think just wraps the `cors` package.
5. Creating a `vercel.json` file in my project root with a `"headers"` array, also shown in the [Vercel doc on enabling CORS](https://vercel.com/guides/how-to-enable-cors#enabling-cors-using-vercel.json).

I'm not 100% sure that this CORS issue is what the Unity dev was experiencing locally before the test app was deployed, so if these seem unrelated to you let me know so we can isolate our troubleshooting!

**Reply by sixpolys** [reply, marked as answer]
this seems to happen when you don't add a trailing / to the url

**Reply by ovieokeh** [reply]
It seems I can reproduce this but with a whole different use case.

I have a route that is used as a webhook for the Whatsapp Business API. And just like OP above, by the time the webhook event reaches my app code, the body is stripped empty.

I'm glad I found this issue because I was nearly going crazy thinking I was doing something wrong haha. Would really appreciate a workaround as I'm working on a highly requested feature.

**Comment by ivankarez** [comment]
Hi @jasonwebb! 👋 Did you found any solution/workaround for this problem? I'am having the same Unity -> Next.js setup.

**Reply by jasonwebb** [reply]
Yep, adding a trailing `/` to the end of the URL in Unity "fixed" it. Probably one of the more ridiculous solves I've seen in my career lol. I never did find anything in the Next.js, Vercel, or Unity docs indicating that this is even a thing, but apparently it is. The Unity dev just randomly tried it one day and we moved on. 🤷🏻

**Comment by cmhand** [comment]
Anyone else encountered this issue? I'm encountering it when I move from local testing to production and a trailing slash has not solved it. 

const payload = await request.json() // returns "Unexpected end of JSON"

Changing to 

const payload = await request.text() // returns an empty string

**Reply by sixpolys** [reply]
Are you sure the trailing slash is added to the production endpoints?
It might help to install a debug proxy like [mitmproxy](https://github.com/mitmproxy/mitmproxy) to see what is being sent, including the headers

**Reply by cmhand** [reply]
Absolutely sure.

I did find the issue though. The request being made to my endpoint did not have a "Content-Length" header, and so vercel was stripping the body. 

Getting the external API to add a valid "Content-Length" header fixed it. Hopefully that'll save someone else some time.
